x-pack/legacy/plugins/ml/server/models/data_visualizer/data_visualizer.ts

```diff
@@ -596,23 +700,39 @@ export class DataVisualizer {
       aggs: buildSamplerAggregation(aggs, samplerShardSize),
     };
 
-    const resp = await this.callWithRequest('search', { index, size, body });
+    const resp = await this.context.core.elasticsearch.dataClient.callAsCurrentUser('search', {
+      index,
+      size,
+      body,
+    });
     const aggregations = resp.aggregations;
     const aggsPath = getSamplerAggregationsResponsePath(samplerShardSize);
-    const batchStats = [];
+    const batchStats: StringFieldStats[] = [];
     fields.forEach((field, i) => {
       const safeFieldName = getSafeAggregationName(field.fieldName, i);
-      const stats = {
-        fieldName: field.fieldName,
-        isTopValuesSampled:
-          field.cardinality >= SAMPLER_TOP_TERMS_THRESHOLD || samplerShardSize > 0,
-      };
 
       const topAggsPath = [...aggsPath, `${safeFieldName}_top`];
       if (samplerShardSize < 1 && field.cardinality >= SAMPLER_TOP_TERMS_THRESHOLD) {
         topAggsPath.push('top');
       }
 
+      const topValues: Bucket[] = _.get(aggregations, [...topAggsPath, 'buckets'], []);
+
+      const stats = {
+        fieldName: field.fieldName,
+        isTopValuesSampled:
+          field.cardinality >= SAMPLER_TOP_TERMS_THRESHOLD || samplerShardSize > 0,
+        topValues,
+        topValuesSampleSize: topValues.reduce(
+          (acc, curr) => acc + curr.doc_count,
+          _.get(aggregations, [...topAggsPath, 'sum_other_doc_count'], 0)
+        ),
+        topValuesSamplerShardSize:
+          field.cardinality >= SAMPLER_TOP_TERMS_THRESHOLD
+            ? SAMPLER_TOP_TERMS_SHARD_SIZE
+            : samplerShardSize,
+      };
+
       stats.topValues = _.get(aggregations, [...topAggsPath, 'buckets'], []);
```

Comment:
Are these lines still needed with the lines you've added above?

Reply:
Good catch, deleted